staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller_test.go

```diff
@@ -28,172 +32,597 @@ import (
+		generatedName1 = generateName + "0"
+		basicDevice    = &resourceapi.BasicDevice{
+			Attributes: map[resourceapi.QualifiedName]resourceapi.DeviceAttribute{
+				"new-attribute": {StringValue: ptr.To("value")},
+			},
+		}
+		nodeSelector = &v1.NodeSelector{
+			NodeSelectorTerms: []v1.NodeSelectorTerm{{
+				MatchFields: []v1.NodeSelectorRequirement{{
+					Key:    "name",
+					Values: []string{"node-a"},
+				}},
+			}},
+		}
+		otherNodeSelector = &v1.NodeSelector{
+			NodeSelectorTerms: []v1.NodeSelectorTerm{{
+				MatchFields: []v1.NodeSelectorRequirement{{
+					Key:    "name",
+					Values: []string{"node-b"},
+				}},
+			}},
+		}
 	)
 
 	testCases := map[string]struct {
-		nodeUID                types.UID
+		// nodeUID is empty if not a node-local.
+		nodeUID types.UID
+		// initialObjects is a list of initial resource slices to be used in the test.
 		initialObjects         []runtime.Object
+		initialOtherObjects    []runtime.Object
 		poolName               string
 		inputDriverResources   *DriverResources
 		expectedResourceSlices []resourceapi.ResourceSlice
 	}{
-		"single-resourceslice-existing": {
+		"no-resourceslices-existing-with-no-devices": {
+			nodeUID:        nodeUID,
+			initialObjects: []runtime.Object{},
+			poolName:       poolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					poolName: {
+						Slices: []Slice{{Devices: []resourceapi.Device{}}},
+					},
+				},
+			},
+			expectedResourceSlices: []resourceapi.ResourceSlice{
+				*MakeResourceSlice().Name(generatedName1).GenerateName(generateName).
+					NodeOwnerReferences(ownerName, string(nodeUID)).NodeName(ownerName).
+					Driver(driveName).Devices([]resourceapi.Device{}).
+					Pool(resourceapi.ResourcePool{Name: poolName, Generation: 1, ResourceSliceCount: 1}).Obj(),
+			},
+		},
+		"no-resourceslices-existing-with-one-device": {
+			nodeUID:        nodeUID,
+			initialObjects: []runtime.Object{},
+			poolName:       poolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					poolName: {
+						Generation: 1,
+						Slices: []Slice{{
+							Devices: []resourceapi.Device{
+								{
+									Name: deviceName,
+								},
+							},
+						}},
+					},
+				},
+			},
+			expectedResourceSlices: []resourceapi.ResourceSlice{
+				*MakeResourceSlice().Name(generatedName1).GenerateName(generateName).
+					NodeOwnerReferences(ownerName, string(nodeUID)).NodeName(ownerName).
+					Driver(driveName).Devices([]resourceapi.Device{{Name: deviceName}}).
+					Pool(resourceapi.ResourcePool{Name: poolName, Generation: 1, ResourceSliceCount: 1}).Obj(),
+			},
+		},
+		"no-resourceslice-existing-with-mismatching-resource-pool-name": {
+			nodeUID:        nodeUID,
+			initialObjects: []runtime.Object{},
+			poolName:       poolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					poolName1: {
+						Generation: 1,
+						Slices: []Slice{{
+							Devices: []resourceapi.Device{
```

Comment:
Do we need a case that covers the scenario where the Devices list in the pool Slice contains multiple Devices?

Reply:
https://github.com/kubernetes/kubernetes/compare/a481793493efdf4267f65c724d7aed71d6fed0e1..0d03fb153e5ef5928e1de18e1ec096c0932c734a